File: rust/tests/budget.rs
//! Offline tests for reading a compute budget across all three versions.

use txv1::budget::{default_compute_unit_limit, ComputeBudget};
use yellowstone_grpc_proto::solana::storage::confirmed_block::{
    CompiledInstruction, Message, TransactionConfig,
};

fn compute_budget_instruction(discriminant: u8, operand: &[u8]) -> CompiledInstruction {
    let mut data = vec![discriminant];
    data.extend_from_slice(operand);
    CompiledInstruction {
        program_id_index: 0,
        accounts: Vec::new(),
        data,
    }
}

/// A legacy or v0 message whose only account key is the ComputeBudget program.
fn instruction_message(instructions: Vec<CompiledInstruction>, versioned: bool) -> Message {
    Message {
        account_keys: vec![solana_sdk_ids::compute_budget::ID.to_bytes().to_vec()],
        config: None,
        instructions,
        versioned,
        ..Default::default()
    }
}

#[test]
fn a_v1_budget_is_read_straight_off_the_config() {
    let message = Message {
        config: Some(TransactionConfig {
            compute_unit_limit: Some(20_000),
            heap_size: Some(65_536),
            loaded_accounts_data_size_limit: Some(65_536),
            priority_fee: Some(5_000),
        }),
        versioned: true,
        ..Default::default()
    };

    let budget = ComputeBudget::of_message(&message);

    assert_eq!(budget.compute_unit_limit, Some(20_000));
    assert_eq!(budget.heap_size, Some(65_536));
    assert_eq!(budget.loaded_accounts_data_size_limit, Some(65_536));
    assert_eq!(budget.priority_fee_lamports, Some(5_000));
}

#[test]
fn a_v0_budget_is_recovered_from_compute_budget_instructions() {
    let message = instruction_message(
        vec![
            compute_budget_instruction(2, &20_000u32.to_le_bytes()),
            compute_budget_instruction(4, &65_536u32.to_le_bytes()),
            compute_budget_instruction(1, &65_536u32.to_le_bytes()),
        ],
        true,
    );

    let budget = ComputeBudget::of_message(&message);

    assert_eq!(budget.compute_unit_limit, Some(20_000));
    assert_eq!(budget.loaded_accounts_data_size_limit, Some(65_536));
    assert_eq!(budget.heap_size, Some(65_536));
}

#[test]
fn a_v0_priority_fee_is_converted_from_micro_lamports_per_unit_to_a_total() {
    // 20,000 CU at 250,000 micro-lamports/CU = 5,000,000,000 micro-lamports,
    // which is the 5,000 lamports a v1 transaction would state directly.
    let message = instruction_message(
        vec![
            compute_budget_instruction(2, &20_000u32.to_le_bytes()),
            compute_budget_instruction(3, &250_000u64.to_le_bytes()),
        ],
        true,
    );

    assert_eq!(
        ComputeBudget::of_message(&message).priority_fee_lamports,
        Some(5_000)
    );
}

#[test]
fn a_v0_priority_fee_rounds_up_to_whole_lamports() {
    // 1 CU at 1 micro-lamport is a millionth of a lamport, which the runtime
    // charges as one lamport rather than as zero.
    let message = instruction_message(
        vec![
            compute_budget_instruction(2, &1u32.to_le_bytes()),
            compute_budget_instruction(3, &1u64.to_le_bytes()),
        ],
        true,
    );

    assert_eq!(
        ComputeBudget::of_message(&message).priority_fee_lamports,
        Some(1)
    );
}

#[test]
fn a_v0_price_without_an_explicit_limit_uses_the_implicit_default() {
    let message = instruction_message(
        vec![compute_budget_instruction(3, &1_000_000u64.to_le_bytes())],
        true,
    );

    // One instruction, no SetComputeUnitLimit, so the limit is the 200k default
    // and a price of one lamport per CU totals 200,000 lamports.
    assert_eq!(default_compute_unit_limit(1), 200_000);
    assert_eq!(
        ComputeBudget::of_message(&message).priority_fee_lamports,
        Some(200_000)
    );
}

#[test]
fn the_implicit_limit_is_clamped_to_the_maximum() {
    assert_eq!(default_compute_unit_limit(7), 1_400_000);
    assert_eq!(default_compute_unit_limit(100), 1_400_000);
}

#[test]
fn a_transaction_with_no_budget_instructions_reports_nothing_set() {
    assert_eq!(
        ComputeBudget::of_message(&instruction_message(Vec::new(), false)),
        ComputeBudget::default()
    );
}

#[test]
fn instructions_from_other_programs_are_ignored() {
    let mut message = instruction_message(
        vec![compute_budget_instruction(2, &20_000u32.to_le_bytes())],
        true,
    );
    // Point the instruction at a different program without changing its data.
    message.account_keys[0] = vec![7u8; 32];

    assert_eq!(
        ComputeBudget::of_message(&message).compute_unit_limit,
        None,
        "only the ComputeBudget program's instructions carry a budget"
    );
}

#[test]
fn a_v1_message_never_falls_back_to_scanning_instructions() {
    // A v1 message carrying a ComputeBudget instruction is malformed, but the
    // config is still the only authority: silently preferring the instruction
    // is exactly the mix-up this accessor exists to prevent.
    let mut message = instruction_message(
        vec![compute_budget_instruction(2, &999_999u32.to_le_bytes())],
        true,
    );
    message.config = Some(TransactionConfig {
        compute_unit_limit: Some(20_000),
        ..Default::default()
    });

    assert_eq!(
        ComputeBudget::of_message(&message).compute_unit_limit,
        Some(20_000)
    );
}

File: go/txv1/budget.go
// Reading a transaction's compute budget regardless of its version.
//
// Legacy and v0 state their budget in ComputeBudget program instructions, v1 in
// the message config. BudgetOfMessage reads both.

package txv1

import (
	"encoding/binary"
	"math"

	"github.com/solana-foundation/solana-go/v2"
	computebudget "github.com/solana-foundation/solana-go/v2/programs/compute-budget"

	"github.com/solana-foundation/transaction-v1-examples/go/pb"
)

const (
	// DefaultComputeUnitsPerInstruction is the compute unit limit applied to each
	// instruction when a legacy or v0 transaction does not set one.
	DefaultComputeUnitsPerInstruction uint32 = 200_000
	// MaxComputeUnitLimit is the ceiling a legacy or v0 transaction's implicit
	// compute unit limit is clamped to.
	MaxComputeUnitLimit uint32 = 1_400_000
)

// BudgetOfMessage reads the compute budget out of a message received over gRPC,
// normalised across all three versions.
//
// v1's four config fields are the normalisation target, so the result is a
// solana.TransactionConfig whatever the version. v1 reads `config` directly;
// legacy and v0 scan ComputeBudget program instructions and convert the fee
// from a price in micro-lamports per compute unit to a total in lamports, so
// the two are comparable.
func BudgetOfMessage(message *pb.Message) solana.TransactionConfig {
	if MessageVersionOf(message) == VersionV1 {
		return budgetFromConfig(message.Config)
	}
	return budgetFromInstructions(message)
}

func budgetFromConfig(config *pb.TransactionConfig) solana.TransactionConfig {
	return solana.TransactionConfig{
		ComputeUnitLimit:            config.ComputeUnitLimit,
		HeapSize:                    config.HeapSize,
		LoadedAccountsDataSizeLimit: config.LoadedAccountsDataSizeLimit,
		PriorityFee:                 config.PriorityFee,
	}
}

func budgetFromInstructions(message *pb.Message) solana.TransactionConfig {
	var budget solana.TransactionConfig
	var priceMicroLamportsPerCU *uint64

	for _, instruction := range message.Instructions {
		if !isComputeBudgetProgram(message.AccountKeys, instruction.ProgramIdIndex) {
			continue
		}
		if len(instruction.Data) == 0 {
			continue
		}
		discriminant, operand := instruction.Data[0], instruction.Data[1:]
		switch discriminant {
		case computebudget.Instruction_RequestHeapFrame:
			budget.HeapSize = readUint32(operand)
		case computebudget.Instruction_SetComputeUnitLimit:
			budget.ComputeUnitLimit = readUint32(operand)
		case computebudget.Instruction_SetComputeUnitPrice:
			priceMicroLamportsPerCU = readUint64(operand)
		case computebudget.Instruction_SetLoadedAccountsDataSizeLimit:
			budget.LoadedAccountsDataSizeLimit = readUint32(operand)
		}
	}

	if priceMicroLamportsPerCU != nil {
		limit := DefaultComputeUnitLimit(len(message.Instructions))
		if budget.ComputeUnitLimit != nil {
			limit = *budget.ComputeUnitLimit
		}
		// The runtime rounds the total up to whole lamports.
		total := divCeil(saturatingMul(uint64(limit), *priceMicroLamportsPerCU), 1_000_000)
		budget.PriorityFee = &total
	}

	return budget
}

// DefaultComputeUnitLimit is the compute unit limit a legacy or v0 transaction
// gets when it sets none.
func DefaultComputeUnitLimit(instructionCount int) uint32 {
	if instructionCount <= 0 {
		return 0
	}
	requested := uint64(DefaultComputeUnitsPerInstruction) * uint64(instructionCount)
	if requested > uint64(MaxComputeUnitLimit) {
		return MaxComputeUnitLimit
	}
	return uint32(requested)
}

func saturatingMul(a, b uint64) uint64 {
	if a == 0 || b == 0 {
		return 0
	}
	if product := a * b; product/a == b {
		return product
	}
	return math.MaxUint64
}

func divCeil(value, divisor uint64) uint64 {
	quotient := value / divisor
	if value%divisor != 0 {
		quotient++
	}
	return quotient
}

// isComputeBudgetProgram reports whether a compiled instruction's program
// account key is the ComputeBudget program.
//
// gRPC delivers account keys as raw protobuf bytes, so any length other than 32
// cannot be an address.
func isComputeBudgetProgram(accountKeys [][]byte, programIDIndex uint32) bool {
	if programIDIndex >= uint32(len(accountKeys)) {
		return false
	}
	key := accountKeys[programIDIndex]
	if len(key) != solana.PublicKeyLength {
		return false
	}
	return solana.PublicKeyFromBytes(key).Equals(solana.ComputeBudget)
}

func readUint32(operand []byte) *uint32 {
	if len(operand) < 4 {
		return nil
	}
	value := binary.LittleEndian.Uint32(operand[:4])
	return &value
}

func readUint64(operand []byte) *uint64 {
	if len(operand) < 8 {
		return nil
	}
	value := binary.LittleEndian.Uint64(operand[:8])
	return &value
}

// ConfigsEqual reports whether two configs set the same fields to the same
// values.
//
// TransactionConfig holds pointers, so `==` compares addresses rather than
// values; and an absent field is not an explicit zero, which is exactly the
// distinction a wire round trip has to preserve.
func ConfigsEqual(a, b solana.TransactionConfig) bool {
	return equalUint64(a.PriorityFee, b.PriorityFee) &&
		equalUint32(a.ComputeUnitLimit, b.ComputeUnitLimit) &&
		equalUint32(a.LoadedAccountsDataSizeLimit, b.LoadedAccountsDataSizeLimit) &&
		equalUint32(a.HeapSize, b.HeapSize)
}

func equalUint32(a, b *uint32) bool {
	if a == nil || b == nil {
		return a == b
	}
	return *a == *b
}

func equalUint64(a, b *uint64) bool {
	if a == nil || b == nil {
		return a == b
	}
	return *a == *b
}
